### PR TITLE
Update storage emulator conformance tests with oauth

### PR DESCRIPTION
**Pull request by tonyjhuang**
### Description

Adds OAuth access token management to storage emulator conformance tests to handle auth handshake when targeting prod.

Also conforms firebase endpoint behavior to prod (GCS endpoint handling still somewhat broken).
- This includes fixing support for `POST /v0/b/<bucket>/o/<object>` (previously the emulator required a `name=<object>` query param

### CODE REVIEW

File: scripts/storage-emulator-integration/conformance/firebase.endpoints.test.ts [hunk 1]
```diff
@@ -66,178 +69,213 @@ describe("Firebase Storage endpoint conformance tests", () => {
...
-  it("should return an error message when uploading a file with invalid metadata", async () => {
-    const fileName = "test_upload.jpg";
-    const errorMessage = await supertest(firebaseHost)
-      .post(`/v0/b/${storageBucket}/o/?name=${fileName}`)
-      .set({ "x-goog-upload-protocol": "multipart", "content-type": "foo" })
-      .expect(400)
-      .then((res) => res.body.error.message);
-
-    expect(errorMessage).to.equal("Invalid Content-Type: foo");
+  describe("multipart upload", () => {
```
Comment (Yuangwang) [inline scripts/storage-emulator-integration/conformance/firebase.endpoints.test.ts:74]: Are most of these changes to fix the conformance tests?

File: src/emulator/storage/apis/firebase.ts [hunk 2]
```diff
@@ -193,16 +193,18 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
...
 
   const handleUpload = async (req: Request, res: Response) => {
-    if (!req.query.name) {
-      res.sendStatus(400);
-      return;
-    }
-
     const bucketId = req.params.bucketId;
-    const objectId = req.query.name.toString();
+    const objectId: string | null = req.params.objectId
```
Comment (Yuangwang) [inline src/emulator/storage/apis/firebase.ts:197]: Could you separate this out into another pr or add some more description to this pr for what this fixes? I'm guessing you caught some discrepancies while running conformance tests but its not too clear

File: scripts/storage-emulator-integration/conformance/firebase.endpoints.test.ts (hunk 1 above)
Comment (tonyjhuang) [reply scripts/storage-emulator-integration/conformance/firebase.endpoints.test.ts:75]: yep, most of these tests were broken when pointed at prod

File: src/emulator/storage/apis/firebase.ts (hunk 2 above)
Comment (tonyjhuang) [reply src/emulator/storage/apis/firebase.ts:197]: Updated the PR description.

When creating a new object, prod supports the following url patterns:
* `v0/b/<bucket>/o/<new object>`
* `v0/b/<bucket>/o?name=<new object>`

This PR fixes the conformance test in `firebase.endpoints.test.ts` to ensure that the emulator supports both

File: src/emulator/storage/multipart.ts [hunk 3]
```diff
@@ -125,11 +125,11 @@ export function parseObjectUploadMultipartRequest(
   body: Buffer
 ): ObjectUploadMultipartData {
   if (!contentTypeHeader.startsWith("multipart/related")) {
-    throw new Error(`Invalid Content-Type: ${contentTypeHeader}`);
+    throw new Error(`Bad content type. ${contentTypeHeader}`);
```
Comment (tonyjhuang) [inline src/emulator/storage/multipart.ts:128]: also addresses broken conformance test in `firebase.endpoints.test.ts`

**Review by Yuangwang — APPROVED** [review]
LGTM
